core/query/sql.ml

```diff
@@ -208,21 +257,21 @@ and string_of_base db one_table b =
       | Apply ("LIKE", [v; w]) -> "(" ^ sb v ^ ")" ^ " LIKE " ^ "(" ^ sb w ^ ")"
       | Apply (f, args) when SqlFuns.is f -> SqlFuns.name f ^ "(" ^ String.concat "," (List.map sb args) ^ ")"
       | Apply (f, args) -> f ^ "(" ^ String.concat "," (List.map sb args) ^ ")"
-      | Empty q -> "not exists (" ^ string_of_query db true q ^ ")"
-      | Length q -> "select count(*) from (" ^ string_of_query db true q ^ ") as " ^ fresh_dummy_var ()
+      | Empty q -> "not exists (" ^ string_of_query quote true q ^ ")"
+      | Length q -> "select count(*) from (" ^ string_of_query quote true q ^ ") as " ^ fresh_dummy_var ()
       | RowNumber [] -> "1"
       | RowNumber ps ->
-        "row_number() over (order by " ^ String.concat "," (List.map (string_of_projection db one_table) ps) ^ ")"
-and string_of_projection db one_table (var, label) =
+        "row_number() over (order by " ^ String.concat "," (List.map (string_of_projection quote one_table) ps) ^ ")"
+and string_of_projection quote one_table (var, label) =
   if one_table then
-    db#quote_field label
+    quote label
   else
-    string_of_table_var var ^ "." ^ (db#quote_field label)
+    string_of_table_var var ^ "." ^ (quote label)
 
-let string_of_query db range q =
+let string_of_query quote range q =
```

Comment:
Possible alternative interface:
```ocaml
val string_of_query : ?quote:(string -> string) -> ?range:(int * int) -> query -> string
```
with defaults
```ocaml
let string_of_query ?(quote=id) ?range q = ...
```

Reply:
Hopefully you'll forgive me for not taking your suggestion here:

  - I'm reluctant to give `quote` a default argument, since the field quotation is something which should almost certainly be set to `db#quote_field` and I fear that having this filled implicitly would open the door to subtle bugs for others later on.

  - Changing `range` to take an `int * int` pair would complicate matters since there's a difference as to whether `range` is specified in a call, or whether `range` is indeed set to `None` at a program level. Conceptually, this would end up with us having an `(int * int) option option` type, which is something I'd like to avoid.

Reply:
You can provide `None` as default for `range` to flatten it. Or alternatively change the `type range = ...` definition.

Reply:
Huh, that works nicely. Thanks.